### ISSUE TITLE
Investigate single-project mode for ad-hoc navigation

### ISSUE THREAD

**Issue by Alexander-Miller** [Alexander-Miller/treemacs#176]
There's demand for it, so I should supply it.

Easiest approach is probably just re-enable uproot & downroot functions when there is only a single project in the workspace, guarded by explicit minor-mode activation to avoid confusion.

**Comment by fiveNinePlusR** [comment]
What are your thoughts on moving toward a model like you have now but with a few modifications?

An ideal way from my perspective would be to have it have one "project" that was arbitrary root navigation at the top that is isolated per frame like it was in the pre-project current treemacs... or perhaps with a toggle from the projects view to the ad-hoc navigation view if that's not feasible.   

And, in addition, modify the current treemacs to have bookmarks like it is now but to allow for setting arbitrary root views in the tree. for instance pressing `h` and `l` to expand and limit the view inside one bookmark but not allow `h` to move up in the folder hierarchy past where the initial bookmark was set.

Example: 

bookmark set at `/foo/bar/` with a subfolder of `baz`. pressing `l` on baz would only show the contents of `baz` while if you were at the folder `bar` and pressed `h` it would not do anything.   

Is there a fundamental reason for the bookmark nesting not being allowed that couldn't be worked around? 

The reason I ask is that say you have the mozilla codebase that is huge with tons of folders that are deeply nested. You want the main folder to have a bookmark but you also want to bookmark a few subfolders feature `a` requires the majority of the time being spent inside folders `a/b/c/d/e` and say `z/x/y/w/v/u/t` that you may be working on for a short time. 

I would like to try to help out with making this a reality if it's possible.

**Comment by Alexander-Miller** [comment]
> modify the current treemacs to have bookmarks like it is now but to allow for setting arbitrary root views in the tree

Bookmarks aren't the right place for this sort of thing, they're just a simple bridge to the builtin bookmark system. Projects are the leading metaphor.

The problem with arbitrary movement is that every directory may only appear at most once as a treemacs node. This works fine when you display discrete project, but you cannot just decide to get a top down view of things and move one project's location to "/". 

It's an annoying conflict, so I have removed root movement altogether, at least at first. Drilling down, as you say, is still useful and possible with the current setup. It's just that you must be prevented from moving beyond the original project root when moving back up. In addition a project should somehow be marked as not showing the full picture. I suppose a new face for the root will do. Or maybe display it as something like **Project Name**<sub>(detailed view)</sub>?

That's one part of the solution. We also need to decide when to allow complete arbitrary root movement. I think that the best answer would be to allow it when there is only a single project in the workspace in the project since interference between projects is then not possible. Anything else, like allowing movement so long as the projects do not intersect is just a pandora's box of complexity.

> I would like to try to help out with making this a reality if it's possible.

You can help by starting to bring back the movement functions. They are still around in v1.18. Some things to consider:
 * Better names would be good. I was never quite happy with `treemacs-uproot` and whatever the other is called.
 * Both functions would call in code that makes sure that backing data structures are still correctly aligned after the root change. That code is now 1) probably no longer correct, and 2) with the feature being much more marginal no longer needed (it's mostly about treemacs remembering which dir is open where). That stuff should be left out.
 * Uproot must of course check for intersections.

**Comment by fiveNinePlusR** [comment]
> every directory may only appear at most once as a treemacs node

Can you expand on why overlapping projects are not allowed/why every node must appear only once?

What are the fundamental data structures at play here that make this a problem?

**Comment by Alexander-Miller** [comment]
It's mostly technical, but also conceptual reasons. 

I call my backing data structure the shadow tree, because it (mostly) shadows the visible file tree in treemacs (I guess it's kind of like a dom and maybe I should call it that, but I don't know enough about the frontend to make that comparison). You can read the gory details in #143. Long story short it's a unified way to track what's open, where things are in the buffer, and what needs refreshing.

Every node in the shadow tree has a unique natural key in its absolute path. That key is for example used as a hashkey to access every node. and data like its position, in constant time. Without the path being unique I guess I'd have to start reference counting, and I'm not sure that system could actually work.

For example you know how when you expand /foo and foo/bar/, close and reopen /foo/ treemacs will also open /bar/ for you because it remembers you had it expanded? The way it works is that /foo/ is not removed from the shadow tree on closing. Because it had expanded child nodes it's merely marked as closed, so its children can be reopened when /foo/ is next expanded. When /foo/ appears twice in treemacs, but you expanded /foo/bar/ on only one instance, how is treemacs to know when to expand /bar/ when you open either of these?

Then there's the issue of follow-mode. How should it handle moving to a file multiple times? What about the internal goto-button-at-that-path function? How should filewatch mode decide which directories to refresh (the current implementation looks for the "highest" dirs needing an update and basically just presses TAB twice on them).

**Comment by fiveNinePlusR** [comment]
Thanks for the thorough response. 

> Every node in the shadow tree has a unique natural key in its absolute path.

Instead of a natural key couldn't you use a projectname(maybe a uuid you generate for the project and that is stored for it)+path as the key?

> When /foo/ appears twice in treemacs, but you expanded /foo/bar/ on only one instance, how is treemacs to know when to expand /bar/ when you open either of these?

The above composite key would solve that too. 

The follow-mode would be a little bit trickier but I am sure it's not insurmountable if it's a direction worth pursuing.

**Comment by Alexander-Miller** [comment]
> maybe a uuid you generate for the project and that is stored for it)+path as the key?

I thought of that too as I was typing this up. I think it could work. I'd have to investigate, make sure the rest of my framework doesn't break down because of some stupid detail.

Still it'd also be a major overhaul, and I'd like to keep my distance from these for a while and work off all the other smaller issues first.

**Comment by Alexander-Miller** [comment]
I've now re-introduced a basic form a free navigation. h & l will allow free root changing so long as there is only a single project in the workspace. It lacks some finer details, like remembering which files were open, and it's not documented yet. I'll wait for feedback before finalising this.
